// @workos-inc/node/lib/pipes/pipes.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jest_fetch_mock_1 = __importDefault(require("jest-fetch-mock"));
const test_utils_1 = require("../common/utils/test-utils");
const workos_1 = require("../workos");
const get_access_token_success_json_1 = __importDefault(require("./fixtures/get-access-token-success.json"));
const get_access_token_no_expiry_json_1 = __importDefault(require("./fixtures/get-access-token-no-expiry.json"));
const get_access_token_not_installed_json_1 = __importDefault(require("./fixtures/get-access-token-not-installed.json"));
const get_access_token_needs_reauth_json_1 = __importDefault(require("./fixtures/get-access-token-needs-reauth.json"));
describe('Pipes', () => {
    let workos;
    beforeAll(() => {
        workos = new workos_1.WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU', {
            apiHostname: 'api.workos.test',
            clientId: 'proj_123',
        });
    });
    beforeEach(() => jest_fetch_mock_1.default.resetMocks());
    describe('getAccessToken', () => {
        it('returns access token with expiry date', () => __awaiter(void 0, void 0, void 0, function* () {
            (0, test_utils_1.fetchOnce)(get_access_token_success_json_1.default);
            const response = yield workos.pipes.getAccessToken({
                provider: 'test-provider',
                userId: 'user_123',
                organizationId: 'org_456',
            });
            expect((0, test_utils_1.fetchURL)()).toContain('/data-integrations/test-provider/token');
            expect((0, test_utils_1.fetchBody)()).toEqual({
                user_id: 'user_123',
                organization_id: 'org_456',
            });
            expect(response).toEqual({
                active: true,
                accessToken: {
                    object: 'access_token',
                    accessToken: 'test_access_token_123',
                    expiresAt: new Date('2025-10-18T12:00:00.000Z'),
                    scopes: ['read:users', 'write:users'],
                    missingScopes: [],
                },
            });
        }));
        it('returns access token without expiry date', () => __awaiter(void 0, void 0, void 0, function* () {
            (0, test_utils_1.fetchOnce)(get_access_token_no_expiry_json_1.default);
            const response = yield workos.pipes.getAccessToken({
                provider: 'test-provider',
                userId: 'user_789',
            });
            expect((0, test_utils_1.fetchURL)()).toContain('/data-integrations/test-provider/token');
            expect((0, test_utils_1.fetchBody)()).toEqual({
                user_id: 'user_789',
                organization_id: undefined,
            });
            expect(response).toEqual({
                active: true,
                accessToken: {
                    object: 'access_token',
                    accessToken: 'test_access_token_456',
                    expiresAt: null,
                    scopes: ['read:data'],
                    missingScopes: ['write:data'],
                },
            });
        }));
        it('returns not_installed failure when integration is not installed', () => __awaiter(void 0, void 0, void 0, function* () {
            (0, test_utils_1.fetchOnce)(get_access_token_not_installed_json_1.default);
            const response = yield workos.pipes.getAccessToken({
                provider: 'test-provider',
                userId: 'user_123',
            });
            expect((0, test_utils_1.fetchURL)()).toContain('/data-integrations/test-provider/token');
            expect(response).toEqual({
                active: false,
                error: 'not_installed',
            });
        }));
        it('returns needs_reauthorization failure when token needs refresh', () => __awaiter(void 0, void 0, void 0, function* () {
            (0, test_utils_1.fetchOnce)(get_access_token_needs_reauth_json_1.default);
            const response = yield workos.pipes.getAccessToken({
                provider: 'test-provider',
                userId: 'user_123',
            });
            expect((0, test_utils_1.fetchURL)()).toContain('/data-integrations/test-provider/token');
            expect(response).toEqual({
                active: false,
                error: 'needs_reauthorization',
            });
        }));
        it('throws error for server errors', () => __awaiter(void 0, void 0, void 0, function* () {
            (0, test_utils_1.fetchOnce)({ message: 'Internal Server Error' }, { status: 500 });
            yield expect(workos.pipes.getAccessToken({
                provider: 'test-provider',
                userId: 'user_123',
            })).rejects.toThrow();
        }));
    });
});
